pynapple/process/signal_processing.py

```diff
@@ -0,0 +1,267 @@
+        If list, define the frequency range, as [freq_start, freq_stop, freq_step].
+        The `freq_step` is optional, and defaults to 1. Range is inclusive of `freq_stop` value.
+    n_cycles : float or 1d array
+        Length of the filter, as the number of cycles for each frequency.
+        If 1d array, this defines n_cycles for each frequency.
+    scaling : float
+        Scaling factor.
+    norm : {'sss', 'amp'}, optional
+        Normalization method:
+
+        * 'sss' - divide by the square root of the sum of squares
+        * 'amp' - divide by the sum of amplitudes
+
+    Returns
+    -------
+    mwt : 2d array
+        Time frequency representation of the input signal.
+
+    Notes
+    -----
+    This computes the continuous wavelet transform at specified frequencies across time.
+    """
+    if not isinstance(sig, (nap.Tsd, nap.TsdFrame, nap.TsdTensor)):
+        raise TypeError("`sig` must be instance of Tsd, TsdFrame, or TsdTensor")
+    if isinstance(freqs, (tuple, list)):
+        freqs = _create_freqs(*freqs)
+    if fs is None:
+        fs = sig.index.shape[0] / (sig.index.max() - sig.index.min())
+    n_cycles = _check_n_cycles(n_cycles, len(freqs))
+    if isinstance(sig, nap.Tsd):
+        sig = sig.reshape((sig.shape[0], 1))
+        output_shape = (sig.shape[0], len(freqs))
+    else:
+        output_shape = (sig.shape[0], len(freqs), *sig.shape[1:])
+        sig = sig.reshape((sig.shape[0], np.prod(sig.shape[1:])))
+    mwt = np.zeros(
+        [sig.values.shape[0], len(freqs), sig.values.shape[1]], dtype=complex
+    )
+    for channel_i in range(sig.values.shape[1]):
+        for ind, (freq, n_cycle) in enumerate(zip(freqs, n_cycles)):
+            mwt[:, ind, channel_i] = _convolve_wavelet(
+                sig[:, channel_i], fs, freq, n_cycle, scaling, norm=norm
+            )
+    if len(output_shape) == 2:
+        return nap.TsdFrame(
+            t=sig.index, d=mwt.reshape(output_shape), time_support=sig.time_support
+        )
+    return nap.TsdTensor(
+        t=sig.index, d=mwt.reshape(output_shape), time_support=sig.time_support
+    )
+
+
+def _convolve_wavelet(
+    sig, fs, freq, n_cycles=1.5, scaling=1.0, precision=10, norm="sss"
+):
+    """
+    Convolve a signal with a complex wavelet.
+
+    Parameters
+    ----------
+    sig : pynapple.Tsd
+        Time series to filter.
+    fs : float
+        Sampling rate, in Hz.
+    freq : float
+        Center frequency of bandpass filter.
+    n_cycles : float, optional, default: 7
+        Length of the filter, as the number of cycles of the oscillation with specified frequency.
+    scaling : float, optional, default: 0.5
+        Scaling factor for the morlet wavelet.
+    norm : {'sss', 'amp'}, optional
+        Normalization method:
+
+        * 'sss' - divide by the square root of the sum of squares
+        * 'amp' - divide by the sum of amplitudes
+
+    Returns
+    -------
+    array
+        Complex- valued time series.
```

Comment:
I would refactor this code to have:
1.  a function that creates the filter banks of all wavelets for the specified parameters and stores it in an NDArray.
2. a function that computes the normalization factor for different method
3. a function that calls the one that creates the filter bank, the normalization factor, uses `nap.convolve` over the filter bank and normalizes the output. 

The modularity of this approach means that if in the future want a new set of wavelets all you need to do is define a new filter bank creating function. On the wavelet class, you can now pass a simple string saying which wavelet you want to compute and it will call the appropriate filter bank generation.

Reply:
I agree but that could potentialy lead to a very complicated way to do a tfd in pynapple